pkg/kv/kvnemesis/generator.go

```diff
@@ -121,7 +179,7 @@ func NewDefaultConfig() GeneratorConfig {
 	// nontransactional batch are disjoint and upgrading to a transactional batch
 	// (see CrossRangeTxnWrapperSender) if they are. roachpb.SpanGroup can be used
 	// to efficiently check this.
-	config.OpPs[OpPBatch] = 0
+	config.Ops.Batch = BatchOperationConfig{}
```

Comment:
Okay so this is wild. I tried making this into `config.Ops.Batch.Ops.PutExisting = 0` instead, but it was failing assertions. The following batch committed at 2 different timestamps!? This seems to be a different case than the comment is describing.

```
  {
    b := &Batch{}
    b.Put(ctx, /Table/50/"c1556caf", v-5) // nil
    b.Get(ctx, /Table/50/"43275686") // (nil, nil)
    b.Put(ctx, /Table/50/"0998fa3b", v-6) // nil
    db0.Run(ctx, b) // nil
  }
```
```
committed batch different timestamps: /Table/50/"0998fa3b":1583976139.381769000,0->v-6 /Table/50/"c1556caf":1583976139.381491000,0->v-5
```

This seems to go away if I also disallow Get in the batches.

Reply:
Were the two writes to different ranges?

Reply:
I'm able to reproduce and do see the requests going to the same range, but it looks like they somehow ended up in their own batch or something? More investigation needed.

Reply:
Oh, I see, we're just straight-up missing a check that non-transactional batches don't get split into sub-batches due to request compatibility rules (as opposed to range boundaries). We'd want something like:
```
diff --git a/pkg/kv/kvclient/kvcoord/dist_sender.go b/pkg/kv/kvclient/kvcoord/dist_sender.go
index 36954d9867..9ff2620ddd 100644
--- a/pkg/kv/kvclient/kvcoord/dist_sender.go
+++ b/pkg/kv/kvclient/kvcoord/dist_sender.go
@@ -692,6 +693,12 @@ func (ds *DistSender) Send(
                splitET = true
        }
        parts := splitBatchAndCheckForRefreshSpans(ba, splitET)
+       if len(parts) > 1 && ba.Txn == nil && ba.IsTransactional() && ba.ReadConsistency == roachpb.CONSISTENT {
+               // If there's no transaction and ba spans ranges, possibly re-run as part of
+               // a transaction for consistency. The case where we don't need to re-run is
+               // if the read consistency is not required.
+               return nil, roachpb.NewError(&roachpb.OpRequiresTxnError{})
+       }
        if len(parts) > 1 && (ba.MaxSpanRequestKeys != 0 || ba.TargetBytes != 0) {
                // We already verified above that the batch contains only scan requests of the same type.
                // Such a batch should never need splitting.
```
This seems like such a large oversight that I wonder if there's something I'm missing about how this is expected to work. This is all super old code that hasn't been considered in a while. @tbg do you know how we could have missed this?

Reply:
Also, I occasionally hit `missing lock at sequence` after that patch. This is being fixed in https://github.com/cockroachdb/cockroach/pull/46019 in case you also run into it.

Reply:
> @tbg do you know how we could have missed this?

Easy enough, we don't ever use such batches ourselves and much of the non-transactional API has basically been neglected for a long time. My expectation is that we want to remove it, I'd be curious to know if you see any roadblocks to doing so (i.e. nontransactional stuff will always be wrapped in a txn). AFAICT the few legitimate uses of nontransactional operations that we have (id generators etc) will be fine with the 1PC path (perhaps with the "allow only 1PC flag" sprinkled in).